iothub/service/src/Amqp/ClientWebSocketTransport.cs

```diff
@@ -0,0 +1,406 @@
+    {
+        private static readonly AsyncCallback s_onReadComplete = OnReadComplete;
+        private static readonly AsyncCallback s_onWriteComplete = OnWriteComplete;
+        private static readonly TimeSpan s_closeTimeout = TimeSpan.FromSeconds(30);
+
+        // Disposables
+
+        private readonly ClientWebSocket _webSocket;
+        private readonly bool _disposeClientWebSocket;
+        private readonly CancellationTokenSource _writeCancellationTokenSource;
+        private bool _isDisposed;
+
+        internal ClientWebSocketTransport(ClientWebSocket webSocket, bool disposeClientWebSocket)
+            : base("clientwebsocket")
+        {
+            _webSocket = webSocket;
+            _disposeClientWebSocket = disposeClientWebSocket;
+            _writeCancellationTokenSource = new CancellationTokenSource();
+        }
+
+        public override string LocalEndPoint => null; // Unused
+
+        public override string RemoteEndPoint => null; // Unused
+
+        public override bool RequiresCompleteFrames => true;
+
+        public override bool IsSecure => true;
+
+        public override void SetMonitor(ITransportMonitor usageMeter)
+        {
+            // Do Nothing
+        }
+
+        public override bool WriteAsync(TransportAsyncCallbackArgs args)
+        {
+            ThrowIfNotOpen();
+
+            Debug.Assert(args.Buffer != null || args.ByteBufferList != null, "must have a buffer to write");
+            Debug.Assert(args.CompletedCallback != null, "must have a valid callback");
+            args.Exception = null; // null out any exceptions
+
+            Task taskResult = WriteImplAsync(args);
+            if (WriteTaskDone(taskResult, args))
+            {
+                return false;
+            }
+
+            ToAsyncResult(taskResult, s_onWriteComplete, args);
+            return true;
+        }
+
+        public override bool ReadAsync(TransportAsyncCallbackArgs args)
+        {
+            ThrowIfNotOpen();
+
+            // Read with buffer list not supported
+            Debug.Assert(args.Buffer != null, "must have buffer to read");
+            Debug.Assert(args.CompletedCallback != null, "must have a valid callback");
+
+            Argument.ValidateBufferBounds(args.Buffer, args.Offset, args.Count);
+            args.Exception = null; // null out any exceptions
+
+            Task<int> taskResult = ReadImplAsync(args);
+            if (ReadTaskDone(taskResult, args))
+            {
+                return false;
+            }
+
+            ToAsyncResult(taskResult, s_onReadComplete, args);
+            return true;
+        }
+
+        public void Dispose()
+        {
+            if (!_isDisposed)
+            {
+                // A user may opt to not dispose the websocket just because the service client is being
+                // disposed.
+                if (_disposeClientWebSocket)
+                {
+                    _webSocket.Abort();
```

Comment:
What does Abort do, and do we not care about doing it if we don't own it?

Reply:
It cancels any connection opened on that web socket and cancels any pending I/O operations. In the case of a client web socket that we created, it is safe to abort because it wouldn't be involved in any connections other than the AMQP connection that is being disposed here. In the case of a client web socket created by the user, we don't know if any other connections are being fed into that web socket, so we should avoid aborting it.